robotoff/workers/queues.py

```diff
@@ -1,23 +1,58 @@
+import random
 import threading
 import time
 from typing import Callable, Optional
 
 from rq import Queue
 from rq.job import Job
 
+from robotoff import settings
 from robotoff.redis import redis_conn
-from robotoff.types import WorkerQueue
+from robotoff.types import ProductIdentifier
+from robotoff.utils import get_logger
 
-high_queue = Queue(WorkerQueue.robotoff_high.value, connection=redis_conn)
-low_queue = Queue(WorkerQueue.robotoff_low.value, connection=redis_conn)
+logger = get_logger(__name__)
+high_queues = [
+    Queue(f"robotoff-high-{i+1}", connection=redis_conn)
+    for i in range(settings.NUM_RQ_WORKERS)
+]
+low_queue = Queue("robotoff-low", connection=redis_conn)
+
+
+def get_high_queue(product_id: Optional[ProductIdentifier] = None):
+    """Return the high-priority queue that is specific to a product.
+
+    There are as many high priority queues as they are workers.
+    We select one of the queues using the product barcode value.
+    This way, over all possible barcodes, all queues are returned with equal
+    probability, but we make sure we always use the same queue for a single
+    product. We greatly reduce the risk of concurrent processing, DB
+    deadlocks,...
+
+    If `product_id` is None, we return one of the high-priority queue
+    randomly.
+
+    :param product_id: the product identifier
+    :return: the selected queue
+    """
+    if product_id is None:
+        return random.choice(high_queues)
+
+    # All product barcodes should be digits, but just make sure the function still
+    # works with non-digit barcodes
+    queue_idx = (int(product_id.barcode) if product_id.barcode.isdigit() else 0) % len(
+        high_queues
+    )
```

Comment:
do you think % works well here ? Do we have an even distribution of last digits for products code ? Also we might want to support non numerical barcode in the future. So why not hash the barcode before using modulo ?

Reply:
I thought about it, and considered it was a bit overkill to use a hashing function for this. I don't know how even the distribution is though.

Reply:
I would find it more robust. (a md5 or sha1 does not cost a lot)

Reply:
You're right, I guess it was just laziness from my part :smile: 
I added it (+ some unit tests)